### PR TITLE
feat: Add log_prob method and endpoint

### LINKED ISSUE #113: Missing `log_prob` method or endpoint
Given a `model_name` and parameter values, call the `log_prob` function in the C++ model. Implementation can mirror that of the `param_names` endpoint.

When this is implemented, add a method to the PyStan 3 `Model` class.

(I think I have used this function one time in the last 4 years.)

### PR DESCRIPTION
**Pull request by mjcarter95**
Given a model name, associated data and unconstrained parameters,
the log_prob endpoint will calculate and return the log probability
of the unconstrained parameters.

This feature is accompanied by two tests, both of which are based on
a simple Gaussian problem: the first test validates the log_prob endpoint
by comparing the output against the analytically derived log probability;
the second test validates the log_prob endpoint by comparing the output
against the log probability (lp__) extracted from a model fit.

Closes #113

### CODE REVIEW

File: httpstan/stan_services.cpp
Comment (ahartikainen) [inline httpstan/stan_services.cpp:208]: adjust_transform

File: tests/test_log_prob.py
Comment (ahartikainen) [inline tests/test_log_prob.py:18]: should this be pytest parameter?

**Review by ahartikainen — COMMENTED** [review]
Should there also be a test against normal model execution (extract lp__ and values -> calc lp with log_prob and compare)?

Maybe with a model containing restriction (e.g. <lower=0>).

**Comment by mjcarter95** [comment]
I've addressed points 2 and 3 in the latest commit.

On the additional test: I think that adding a restriction would require `transform_inits` to be available to transform from the constrained to unconstrained space - please let me know if that's not correct. If that's correct, I can add an additional function to fit the Gaussian model (without constraints) and compare lp__ to log_prob with a view to adding a model with constraints once `transform_inits` is available.

**Comment by ahartikainen** [comment]
I will let @riddell-stan to comment, but I think having functions to constrain / unconstrain parameter values is needed this to be useful.

Also, what order does this use if variables are vector / matrix? Or does the code handle this already?

**Comment by mjcarter95** [comment]
Agreed on the need for functions to constrain/unconstrain parameters, @riddell-stan and I discussed adding more functions in #113 but agreed to start with `log_prob`.

The function accepts a list with length equal to the n umber of unconstrained: I think this mirrors the PyStan implementation, the only difference being that PyStan accepts a numpy array.

**Comment by ahartikainen** [comment]
Using lists is good. We don't need numpy in httpstan.

I also wonder what is the parameter order and if that order is the same with the output we get from Stan.

File: tests/test_log_prob.py
Comment (ahartikainen) [reply tests/test_log_prob.py:19]: https://docs.pytest.org/en/stable/parametrize.html

**Comment by mjcarter95** [comment]
I performed some tests in terminal using eight schools: looks like the order of the inputs is equal to Stan. I'm not sure how we can validate that the user has the order correct in the input list, accepting a dictionary and then building the input list might make this easier?

Latest commit hopefully fixes formatting issues and contains an additional test that fits the Gaussian example and then compares the output from lp__ and log_prob.

File: httpstan/views.py
Comment (riddell-stan) [inline httpstan/views.py:562]:
```suggestion
              Boolean to control whether we apply a Jacobian adjust transform.
```

**Comment by riddell-stan** [comment]
This looks good to me. The mypy errors need to be fixed, of course.

Also, can you rewrite the commit message to reference the issue in the commit body? We haven't mentioned issue numbers in the commit summary.

Something like this:

```
feat: Add log_prob method and endpoint

Given a model name, associated data and unconstrained parameters,
the log_prob endpoint will calculate and return the log probability
of the unconstrained parameters.

This feature is accompanied by a test: the log_prob test validates
the endpoint through a simple Gaussian toy problem.

Closes #113
```

**Comment by mjcarter95** [comment]
Adjusted the commit message. The latest commit also addressed two of the mypy errors, I'm not sure what is causing the remaining error, I'll have chance to review again in ~3 hours.

```
+mypy httpstan tests
httpstan/views.py:590: error: Module has no attribute "log_prob"
Found 1 error in 1 file (checked 47 source files)
```

**Comment by riddell-stan** [comment]
@mjcarter95 sorry to heap one more thing on you. Can you check out https://github.com/stan-dev/rstan/issues/813#issuecomment-665351890 ? We definitely want to avoid confusion here. For example, `log_prob` should not return something different than RStan's `log_prob`. If RStan's `log_prob` doesn't make sense we could perhaps rename this endpoint to something more specific.

If you can't figure out the mypy issue I'll have a look. Let me know. Thanks for working on this.

**Comment by mjcarter95** [comment]
No problem. Some comparisons against PyStan below (I don't have R or RStan installed but it looks like the underlying function calls are the same.  The log_prob endpoint is setup in a similar fashion to RStan/PyStan in that we call `log_prob_propto` and users can turn the Jacobian adjust transform on/off and the output is the same as PyStan (I assume RStan as well) when `adjust_transform` is set to True or False.

```
Gaussian (adjust_transform=True, adjust_transform=False)
x: [1]
log_prob endpoint:
log_prob PyStan: -0.5, -0.5

Eight Schools
x: [0 1 2 3 4 5 6 7 8 9]
log_prob endpoint: -146.18938914140082, -147.18938914140082
log_prob PyStan: -146.18938914140082, -147.18938914140082

Neals Funnel
x: [0 1 2 3 4 5 6 7 8 9]
log_prob endpoint: -142.5, -142.5
log_prob PyStan: -142.5, -142.5
```

**Comment by riddell-stan** [comment]
Thanks for checking.

File: httpstan/views.py
Comment (ahartikainen) [inline httpstan/views.py:590]:
```suggestion
        lp = services_module.log_prob(data, unconstrained_parameters, adjust_transform)  # type: ignore
```

**Review by ahartikainen — COMMENTED** [review]
We can ignore the type here

**Comment by mjcarter95** [comment]
Thank you @ahartikainen looks like that solved the lingering mypy error. Out of interest: how long do the tests normally run for?

**Comment by ahartikainen** [comment]
Tests don't usually go too long. Currently Travis is out of use, so we need to wait until #498 is merged.

**Comment by riddell-stan** [comment]
@mjcarter95 Would you mind rebasing on `main` so the tests will on all the supported platforms? Sorry about this. We didn't anticipate that Travis CI would kick everyone out. :wink:

File: httpstan/stan_services.cpp
Comment (riddell-stan) [inline httpstan/stan_services.cpp:209]: Shouldn't we change this docstring?

We're calling `stan::model::log_prob_propto` with the model in this case, right? And the model *does* have a `log_prob` method (which we are not using, I believe).

**Comment by mjcarter95** [comment]
@riddell-stan I have rebased onto main and updated the docstring as suggested.

**Comment by mjcarter95** [comment]
The tests failed when validating the docstring that I wrote for `handle_log_prob`. I don't have much experience on this front, and the error message is a little unclear to me, any suggestions on what might be wrong would be appreciated.

File: httpstan/views.py
Comment (ahartikainen) [inline httpstan/views.py:606]:
```suggestion
        "200":
          description: Log probability of the unconstrained parameters.
          schema:
            type: object
            properties:
              log_prob:
                type: number
                items: ...
```

I think the error comes from here.

Should it return an object? I'm not sure what should go in `items` if anything?

**Review by ahartikainen — COMMENTED** [review]
Not sure what is the problem, I added one suggestion.

File: httpstan/views.py
Comment (mjcarter95) [reply httpstan/views.py:606]: @ahartikainen Thank you for the suggestions, I amended the first change. On this one, I don't think we need "items" given that the function returns a single value, however I now see the error below

```
Traceback (most recent call last):
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 236, in _validate_openapi_spec_validator
    validate_v2_spec(self.specification)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/openapi_spec_validator/shortcuts.py", line 7, in validate
    return validator_callable(spec, spec_url=spec_url)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/openapi_spec_validator/validators.py", line 48, in validate
    raise err
openapi_spec_validator.exceptions.OpenAPIValidationError: {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas

Failed validating 'oneOf' in schema['properties']['paths']['patternProperties']['^/']['properties']['post']['properties']['responses']['patternProperties']['^([0-9]{3})$|^(default)$']:
    {'oneOf': [{'$ref': '#/definitions/response'},
               {'$ref': '#/definitions/jsonReference'}]}

On instance['paths']['/v1/models/{model_id}/log_prob']['post']['responses']['200']:
    {'description': 'Log probability of the unconstrained parameters.',
     'properties': {'log_prob': {'type': 'number'}},
     'type': 'object'}

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/apispec/utils.py", line 69, in validate_spec
    prance.BaseParser(spec_string=json.dumps(spec.to_dict()), **parser_kwargs)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 107, in __init__
    self.parse()
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 138, in parse
    self._validate()
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 173, in _validate
    validator(parsed)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 240, in _validate_openapi_spec_validator
    raise_from(ValidationError, v2_ex)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/util/exceptions.py", line 25, in raise_from
    raise klass(*args) from from_value
prance.ValidationError: ("{'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas", 'oneOf', deque(['paths', '/v1/models/{model_id}/log_prob', 'post', 'responses', '200']), None, [<ValidationError: "'properties', 'type' do not match any of the regexes: '^x-'">, <ValidationError: "'$ref' is a required property">, <ValidationError: "Additional properties are not allowed ('type', 'properties', 'description' were unexpected)">], [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}], {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}}, {'oneOf': [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}]}, deque(['properties', 'paths', 'patternProperties', '^/', 'properties', 'post', 'properties', 'responses', 'patternProperties', '^([0-9]{3})$|^(default)$', 'oneOf']), None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/home/sgmcart3/Code/stan/httpstan/doc/conf.py", line 62, in make_openapi_spec
    fh.write(openapi.openapi_spec().to_yaml())
  File "/home/sgmcart3/Code/stan/httpstan/httpstan/openapi.py", line 50, in openapi_spec
    apispec.utils.validate_spec(spec)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/apispec/utils.py", line 71, in validate_spec
    raise exceptions.OpenAPIError(*err.args)
apispec.exceptions.OpenAPIError: ("{'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas", 'oneOf', deque(['paths', '/v1/models/{model_id}/log_prob', 'post', 'responses', '200']), None, [<ValidationError: "'properties', 'type' do not match any of the regexes: '^x-'">, <ValidationError: "'$ref' is a required property">, <ValidationError: "Additional properties are not allowed ('type', 'properties', 'description' were unexpected)">], [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}], {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}}, {'oneOf': [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}]}, deque(['properties', 'paths', 'patternProperties', '^/', 'properties', 'post', 'properties', 'responses', 'patternProperties', '^([0-9]{3})$|^(default)$', 'oneOf']), None)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/cmd/build.py", line 279, in build_main
    args.tags, args.verbosity, args.jobs, args.keep_going)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/application.py", line 278, in __init__
    self._init_builder()
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/application.py", line 337, in _init_builder
    self.events.emit('builder-inited')
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/events.py", line 118, in emit
    (listener.handler, name), exc) from exc
sphinx.errors.ExtensionError: Handler <function make_openapi_spec at 0x7f5d4df3b4d0> for event 'builder-inited' threw an exception (exception: ("{'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas", 'oneOf', deque(['paths', '/v1/models/{model_id}/log_prob', 'post', 'responses', '200']), None, [<ValidationError: "'properties', 'type' do not match any of the regexes: '^x-'">, <ValidationError: "'$ref' is a required property">, <ValidationError: "Additional properties are not allowed ('type', 'properties', 'description' were unexpected)">], [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}], {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}}, {'oneOf': [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}]}, deque(['properties', 'paths', 'patternProperties', '^/', 'properties', 'post', 'properties', 'responses', 'patternProperties', '^([0-9]{3})$|^(default)$', 'oneOf']), None))

Extension error:
Handler <function make_openapi_spec at 0x7f5d4df3b4d0> for event 'builder-inited' threw an exception (exception: ("{'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas", 'oneOf', deque(['paths', '/v1/models/{model_id}/log_prob', 'post', 'responses', '200']), None, [<ValidationError: "'properties', 'type' do not match any of the regexes: '^x-'">, <ValidationError: "'$ref' is a required property">, <ValidationError: "Additional properties are not allowed ('type', 'properties', 'description' were unexpected)">], [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}], {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}}, {'oneOf': [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}]}, deque(['properties', 'paths', 'patternProperties', '^/', 'properties', 'post', 'properties', 'responses', 'patternProperties', '^([0-9]{3})$|^(default)$', 'oneOf']), None))
```

Comment (riddell-stan) [reply httpstan/views.py:606]: I wish one didn't have to learn a new "language" (i.e., REST-based APIs) to write code. Alas, it seems like there's no alternative.

I *think* the convention is that you always encapsulate things in a JSON object. I'd have to do more searching to verify this. (It might even be a hard requirement.)

How about ``{"log_prob": value}`` or ``{"log_prob": value}``

Comment (riddell-stan) [reply httpstan/views.py:606]: ``{"data": value}`` might also work. Check out https://google.github.io/styleguide/jsoncstyleguide.xml for ideas. I think that document might also explain the encapsulation issue.

Comment (mjcarter95) [reply httpstan/views.py:606]: Thank you. The sphinx module runs without any errors when removing the following:

```
        - in: body
          name: unconstrained_parameters
          description: >-
              Unconstrained parameters to calculate log probability for.
          required: true
          schema:
            type: array
        - in: body
          name: adjust_transform
          description: >-
              Boolean to control whether we apply a Jacobian adjust transform.
          required: false
          schema:
            type: boolean
```

Comment (mjcarter95) [reply httpstan/views.py:606]: The issue seems to be regarding how I define `type: array`, replacing array with, say, boolean means that sphinx runs with no error.

File: httpstan/views.py
Comment (riddell-stan) [inline httpstan/views.py:570]:
```suggestion
        Returns the output of Stan C++ `stan::model::log_prob_propto`.
```

File: httpstan/views.py
Comment (riddell-stan) [inline httpstan/views.py:571]:
```suggestion
        Returns the output of Stan C++ `stan::model::log_prob_propto`.
```

**Comment by riddell-stan** [comment]
I'm not sure why test coverage went down. Perhaps because we do not use the functions in `views` directly? @ahartikainen Any ideas?

Would be nice to solve this now so it doesn't happen the next time we have a good-looking PR complete with tests.

```
FAIL Required test coverage of 93% not reached. Total coverage: 92.80%
```

**Comment by riddell-stan** [comment]
@mjcarter95 rebase on main and all should work. Sorry about the delay.

This has been a learning experience. With luck future PRs will not require so much work.

**Comment by mjcarter95** [comment]
Thank you @riddell-stan and no problem, all checks pass now.

**Review by riddell-stan — APPROVED** [review]

**Pull request merged**
